src/Casper/Internal.hs

```diff
@@ -0,0 +1,475 @@
+  exists <- liftIO . doesFileExist $ path
+  unless exists $ throwError (MetaMissing path)
+  result <- fmap decodeResourceMeta . liftIO . BS.readFile $ path
+  either (throwError . MetaCorrupt path) pure result
+
+-- | Recall the reference for a SHA256 hash in the store.
+-- If the content is not present in the store 'Nothing' is returned.
+recall :: (Content a, MonadIO m) => SHA256 -> CasperT root m (Maybe (Ref a))
+recall sha = CasperT $ do
+  content <- asks (blobPath sha) >>= liftIO . doesFileExist
+  meta <- asks (contentPath sha) >>= liftIO . doesFileExist
+  if content && meta then pure (Just (Ref sha)) else pure Nothing
+
+-- | Retrieve content from the store using a reference
+retrieve :: (Content a, MonadIO m) => Ref a -> CasperT root m a
+retrieve (Ref sha) = decodeFile (blobPath sha) FileMissing FileCorrupt
+
+blobPath :: SHA256 -> Store root -> FilePath
+blobPath sha s = storeDir s </> "blob" </> show sha
+
+contentPath :: SHA256 -> Store root -> FilePath
+contentPath sha s = storeDir s </> "content" </> show sha
+
+resourcePath :: UUID.UUID -> Store root -> FilePath
+resourcePath uuid s = storeDir s </> "resource" </> show uuid
+
+-- | Store content and obtain a reference to the stored value.
+store ::
+  (Content a, MonadMask m, MonadIO m) =>
+  -- | Value to store
+  a ->
+  -- | Ref to stored value
+  CasperT root m (Ref a)
+store a = withFileLock sha . CasperT $ do
+  exists <- asks (contentPath sha) >>= liftIO . doesFileExist
+  unless exists $ do
+    asks (blobPath sha) >>= writeCasper bs
+    asks (contentPath sha) >>= writeCasper (encodeMeta meta)
+  pure $ Ref sha
+  where
+    writeCasper content path = liftIO $ BS.writeFile path content
+    bs = encodeContent a
+    sha = hashBS bs
+    meta = mkMeta a
+
+withFileLock :: (MonadMask m, MonadIO m) => SHA256 -> CasperT root m a -> CasperT root m a
+withFileLock sha run = do
+  lockVar <- CasperT (asks contentLocks)
+  bracket_ (acquire lockVar) (release lockVar) run
+  where
+    acquire lockVar = liftIO . atomically $ do
+      locks <- readTVar lockVar
+      let isLocked = S.member sha locks
+      when isLocked retry
+      modifyTVar lockVar (S.insert sha)
+    release lockVar =
+      liftIO . atomically $
+        modifyTVar lockVar (S.delete sha)
+
+getRoot :: Monad m => CasperT root m (Loc root)
+getRoot = pure $ Loc UUID.nil
+
+data Store root = Store
+  { storeDir :: FilePath,
+    gcLock :: TVar Bool,
+    contentLocks :: TVar (Set SHA256),
+    resourceLock :: TVar Bool,
+    resourceLocks :: TVar (Set UUID.UUID),
+    activeBlocks :: TVar Natural
+  }
+
+resourcePath' :: UUID.UUID -> FilePath -> FilePath
+resourcePath' uuid dir = dir </> "resource" </> show uuid
+
+initStore :: Resource root => FilePath -> root -> IO (Store root)
+initStore path r = do
+  exists <- doesDirectoryExist path
+  unless exists (createDirectory path)
+  none <- emptyDirectory path
+  when none $ do
+    createDirectory (path </> "blob")
+    createDirectory (path </> "content")
```

Comment:
Maybe `object` would be a better name here but that wouldn't align with the type class.